File: rust-code-examples/src/design_patterns/strategy.ts
// strategy pattern

interface Strategy {
    execute(a: number, b: number): number;
}

class Add implements Strategy {
    execute(a: number, b: number) {
        return a + b;
    }
}

class Subtract implements Strategy {
    execute(a: number, b: number) {
        return a - b;
    }
}

class Context {
    private strategy: Strategy;

    constructor(strategy: Strategy) {
        this.strategy = strategy;
    }

    setStrategy(strategy: Strategy) {
        this.strategy = strategy;
    }

    executeStrategy(a: number, b: number) {
        return this.strategy.execute(a, b);
    }
}

const context = new Context(new Add());
console.log(context.executeStrategy(1, 2));

context.setStrategy(new Subtract());
console.log(context.executeStrategy(1, 2));

File: rust-code-examples/src/design_patterns/strategy.rs
// strategy pattern

trait Strategy {
    fn execute(&self, a: i32, b: i32) -> i32;
}

struct Add;

impl Strategy for Add {
    fn execute(&self, a: i32, b: i32) -> i32 {
        a + b
    }
}

struct Subtract;

impl Strategy for Subtract {
    fn execute(&self, a: i32, b: i32) -> i32 {
        a - b
    }
}

struct Context {
    strategy: Box<dyn Strategy>,
}

impl Context {
    fn new(strategy: Box<dyn Strategy>) -> Self {
        Self { strategy }
    }

    fn set_strategy(&mut self, strategy: Box<dyn Strategy>) {
        self.strategy = strategy;
    }

    fn execute(&self, a: i32, b: i32) -> i32 {
        self.strategy.execute(a, b)
    }
}

pub fn strategy() {
    let context = Context::new(Box::new(Add));
    println!("10 + 5 = {}", context.execute(10, 5));

    let mut context = Context::new(Box::new(Subtract));
    println!("10 - 5 = {}", context.execute(10, 5));

    context.set_strategy(Box::new(Add));
    println!("10 + 5 = {}", context.execute(10, 5));
}
